DEVDOCS-5898: Product reviews (#272)
bigcommerce/docs: docs/api-docs/storefront/graphql/graphql-product-reviews.mdx

# Product Reviews with the GraphQL Storefront API

BigCommerce's GraphQL Storefront API allows you to create and query [product reviews](https://support.bigcommerce.com/s/article/Managing-Reviews) through the built-in comment system. 

## Authentication

To work with product reviews, the [API account](/docs/start/authentication/api-accounts) that generates [Storefront tokens](/docs/rest-authentication/tokens) must include one of the following [OAuth scopes](/docs/start/authentication/api-accounts#oauth-scopes):

| UI Name | Permission | Parameter |
|:--------|:-----------|:----------|

> | Content | modify | `store_v2_content` |

Author: Is this accurate? I tested this OAuth scope briefly

Reviewer: API Account scopes are not relevant to making GQL requests. Once you have a valid Storefront JWT token you will be able to utilize the add product review mutation.

Author: @bc-jz Do you have the answer for this? Or I might remove

Reviewer: I think it should be removed.

Author: Yes, this is the scope needed to generate a JWT. If you know the scope, it would be great if we could include it

Author: Ah I see, I think I'll remove it

| Content | read-only | `store_v2_content_read_only` |

## Store settings

There are a few store settings to consider when creating a product review:

- [Overall enable product reviews](https://support.bigcommerce.com/s/article/Managing-Reviews?language=en_US#enable)
- [Only accept product reviews from past customers](https://support.bigcommerce.com/s/article/Managing-Reviews?language=en_US#enable)
- [Enable reCAPTCHA](https://support.bigcommerce.com/s/article/Updating-Themes-with-reCAPTCHA-v2)

**Overall enable product reviews**

> The merchant must enable this overall setting for the review submission process to work. Disabling this setting cancels any attempts to submit reviews.

Reviewer 2: ```suggestion
The merchant must enable this overall setting for the review submission process. Turning off this setting cancels any attempts to submit reviews.
```

**Only allow reviews from previous customers**

> When the merchant enables the 'Only allow reviews from previous customers' setting, only logged-in customers (not guest shoppers) can submit a review, and they must have at least one previous order placed in the store.

Reviewer: The setting in the CP uses the following verbiage which we should probably match here:

```
Only accept product reviews from past customers
```

Author: fixed, thanks

This setting ensures the merchant has a genuine customer to contact if needed. You can use the [login mutation](/docs/storefront/graphql#customer-login) to log in a customer.

> Once a purchase has been made, customers can review any item in the store. The order doesn't have to include the product being reviewed. 

Reviewer 2: ```suggestion
After a purchase, customers can review any item in the store. The order doesn't have to include the reviewed product. 
```

> When the merchant disables the setting, guest shoppers can also submit reviews, but they must provide an email. Since guests are not logged in, they must supply the email in the request as an input parameter.

Reviewer 2: ```suggestion
When the merchant turns off the setting, guest shoppers can also submit reviews but they must provide an email. Since guests do not log in, they must supply the email in the request as an input parameter.
```

Logged-in customers don't need to provide the email through the input parameter, as the system will use the email from the customer's account.

> If the logged-in customer provides an email as an input, BigCommerce will validate the email against the email associated with the shopper's account. If the email doesn't match, the submission will fail.

Reviewer 2: ```suggestion
If the logged-in customer provides an email as input, BigCommerce will validate the email against the email associated with the shopper's account. If the email doesn't match, the submission will fail.
```

In all cases, you can create only one review per product an email address.

**Enable reCAPTCHA**

If a merchant enables the [reCAPTCHA setting](https://support.bigcommerce.com/s/article/Updating-Themes-with-reCAPTCHA-v2) on the storefront, you need to send a reCAPTCHA token with the request.

## Mutation

The following mutation creates a product review for a store.

<Tabs items={['Request', 'Response']}>
<Tab>

```graphql filename="Example request: Create a product review" showLineNumbers copy
mutation {
  catalog {
    addProductReview (
      reCaptchaV2: "12345"
      input: {
        productEntityId: 81
          review: {
            title: "teitle"
            text: "textr"
            author: "oman"
            rating: 4  # integer value between 1 and 5

>             email: "mem@ail.com2"

Reviewer: Up to you, but maybe use some more "professional" fake data here. I know I gave this to you but I just threw in some random text. Do y'all have any consistent practice for fake data utilized in example input?

```
            title: "review title"
            text: "review body"
            author: "Author Name"
            rating: 4  # integer value between 1 and 5
            email: "emailOfReviewer@myemail.com"
```

Reviewer: Likewise don't forget to update the output if you do change this.

        }
      }
    ){
      errors {
        ...on NotAuthorizedToAddProductReviewError {
          message
        }
        ...on CustomerAlreadyReviewedProductError {
          message
        }
        ...on ProductIdNotFoundError {
          message
        }
        ...on InvalidInputFieldsError {
          message fields
        }
        ...on Error{
          message
        }
      }
    }
  }
}
```

</Tab>
<Tab>

```json filename="Example success response: Create a product review" showLineNumbers copy
{
  "data": {
    "catalog": {
      "addProductReview": {
        "errors": []
      }
    }
  }
}
```

</Tab>
</Tabs>

## Query

> You can query the product review with the `products` graphql node.

Reviewer 2: ```suggestion
You can query the product review with the `products` GraphQL node.
```
